### ISSUE TITLE
Undo tree

### ISSUE THREAD

**Issue by ZehCnaS34** [atom/text-buffer#45]
Just wondering if an undo tree is available in atom.
![undoing](https://cloud.githubusercontent.com/assets/1127553/6129258/e9495922-b10a-11e4-8f60-dedcf0224e17.png)
And if not, could I start implementing a branching history for Undo/Redo?

**Comment by wmayner** [comment]
:+1:

**Comment by mrthebob** [comment]
yes, please!

**Comment by terraqout** [comment]
@ZehCnaS34 Do you start implementing? Can we collaborate?

**Comment by ZehCnaS34** [comment]
No, I didn't. I never got the :+1: at the time. Since atom has reached 1.0, I think it would be ok to start.

but just to make things clear, I kind of want to deviate from the original implementation of the undo tree, and make it more atom*ish*. (hopefully for the better)

Any input?

**Comment by mrthebob** [comment]
Deviate how?

**Comment by ZehCnaS34** [comment]
Really just the user interface.

**Comment by mrthebob** [comment]
Ah, OK.  Cool.  No objections here, then :)

**Comment by terraqout** [comment]
What about persistent history? As I understand it should be implemented from scratch. without it undotree loose half of its usefulness

**Comment by trusktr** [comment]
yes please!!! 👍

**Comment by mkantor** [comment]
Since it's been a while: is this still being worked on? Are there any packages out there that provide this functionality? I searched the repository but was unable to find anything...

**Comment by paradoxxxzero** [comment]
I released a package that tries to implement this behavior (without user interface) see [atom-undo-tree](https://github.com/paradoxxxzero/atom-undo-tree)

**Comment by mrthebob** [comment]
@paradoxxxzero how does it work? Same VIM commands?

**Comment by paradoxxxzero** [comment]
It simply replaces the default undo behavior. I don't know much about the
vim mode but if it uses the default undo command it should work.

**Comment by am** [comment]
@paradoxxxzero I don't think the package fulfils the same behaviour described. The original undo tree concept let's you navigate under a tree, this means you should be able to select different branches at some point. I think the trick here is make history non linear.

**Comment by paradoxxxzero** [comment]
@am It's not the same behavior since there's no UI to navigate the tree, but it's an undo tree that has a default traversal. It's far from being perfect but it could be a stepping stone for anyone willing to give a shot at it.
